chainer/variable.py

```diff
@@ -603,7 +607,7 @@ def _ensure_grad_var_up_to_date(self):
         # _grad_var._data[0] and self._data[0].grad and recreates _grad_var
         # as necessary. (chainerx.ndarray.grad can be altered independently
         # from chainer)
-        if self.xp is chainerx:
+        if self._has_chainerx_array:
```

Comment:
Is this change needed? I think `self.xp` is now fast enough?

Reply:
IMO, we need to do it for two reasons:

1. `xp` is a `property` of the `Variable` which generates further overhead than a function
2. It is safe to refer the independent flag instead of using `array`, `device`  and `xp`, because there is a chance that the attributes are configured as `chainerx` mode but `array` is `None`)

How do you think?

Reply:
>1. `xp` is a `property` of the `Variable` which generates further overhead than a function

Hmm, is it true? In my benchmark, a property is faster than a method and as fast as a global function.
* property/function: ~65ns
* method: ~80ns

>2. It is safe to refer the independent flag instead of using `array`, `device`  and `xp`, because there is a chance that the internal attributes are configured as `chainerx` mode but `array` is `None`)

I think it's a separate issue. We have to reconsider how a variable should behave if the array is `None`.

Reply:
Ah, sorry, you are right. Accessing property is faster than the method call, but accessing attribute is still two or three times faster than other ways.

Anyway, I agree to fix it later as another issue. I'll revert them.

```python
>>> import timeit
>>> setup = """
... class A:
...   def __init__(self):
...     self._f = False
... 
... class B:
...   def f(self):
...     return False
... 
... class C:
...   @property
...   def f(self):
...     return False
... 
... a = A()
... b = B()
... c = C()
... """
>>> timeit.timeit("a._f", setup=setup, number=100000000)
2.5808276534080505
>>> timeit.timeit("b.f()", setup=setup, number=100000000)
8.198761623352766
>>> timeit.timeit("c.f", setup=setup, number=100000000)
7.625560574233532
>>> timeit.timeit("f()", setup="def f(): return False", number=100000000)
5.695897337049246
```

Reply:
>accessing attribute is still two or three times faster than other ways.

Yes, but the difference is only ~40ns (in my benchmark). I doubt it's worthwhile.